File: skills/clawhub/max-self-improvement/scripts/validate_skill.py
#!/usr/bin/env python3
"""
Max-Self-Improvement Skill 验证脚本
验证 SKILL.md frontmatter 格式和目录结构
"""

import os
import re
import sys
from pathlib import Path

def validate_frontmatter(skill_path: Path) -> tuple[bool, list[str]]:
    """验证 frontmatter 格式"""
    errors = []
    skill_md = skill_path / "SKILL.md"
    
    if not skill_md.exists():
        return False, ["SKILL.md not found"]
    
    content = skill_md.read_text(encoding="utf-8")
    
    # 检查是否以 --- 开头
    if not content.startswith("---"):
        errors.append("SKILL.md must start with --- (frontmatter delimiter)")
        return False, errors
    
    # 提取 frontmatter
    frontmatter_match = re.match(r'^---\n(.*?)\n---', content, re.DOTALL)
    if not frontmatter_match:
        errors.append("Cannot parse frontmatter block")
        return False, errors
    
    frontmatter = frontmatter_match.group(1)
    
    # 检查必需字段
    required_fields = ["name", "description"]
    for field in required_fields:
        if not re.search(rf'^{field}:', frontmatter, re.MULTILINE):
            errors.append(f"Missing required field: {field}")
    
    # 检查 name 格式（连字符）
    name_match = re.search(r'^name:\s*(.+)', frontmatter, re.MULTILINE)
    if name_match:
        name = name_match.group(1).strip()
        if not re.match(r'^[a-z0-9-]+$', name):
            errors.append(f"name '{name}' must be lowercase with hyphens only")
    
    return len(errors) == 0, errors


def validate_meta_json(skill_path: Path) -> tuple[bool, list[str]]:
    """验证 _meta.json"""
    import json
    errors = []
    meta_path = skill_path / "_meta.json"
    
    if not meta_path.exists():
        return False, ["_meta.json not found"]
    
    try:
        meta = json.loads(meta_path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as e:
        return False, [f"Invalid JSON: {e}"]
    
    required_fields = ["id", "version"]
    for field in required_fields:
        if field not in meta:
            errors.append(f"Missing required field in _meta.json: {field}")
    
    # 检查 version 格式
    if "version" in meta:
        if not re.match(r'^\d+\.\d+\.\d+$', meta["version"]):
            errors.append("version must be semantic format (e.g., 1.0.0)")
    
    return len(errors) == 0, errors


def validate_structure(skill_path: Path) -> tuple[bool, list[str]]:
    """验证目录结构"""
    errors = []
    
    # 检查必需的目录和文件
    required = ["SKILL.md", "_meta.json"]
    for item in required:
        if not (skill_path / item).exists():
            errors.append(f"Missing required file: {item}")
    
    return len(errors) == 0, errors


def main():
    skill_path = Path(__file__).parent.parent.resolve()
    
    print(f"Validating skill at: {skill_path}")
    print("-" * 50)
    
    all_passed = True
    
    # 1. 结构验证
    passed, errors = validate_structure(skill_path)
    if passed:
        print("[PASS] Structure validation")
    else:
        print("[FAIL] Structure validation")
        for e in errors:
            print(f"  - {e}")
        all_passed = False
    
    # 2. Frontmatter 验证
    passed, errors = validate_frontmatter(skill_path)
    if passed:
        print("[PASS] Frontmatter validation")
    else:
        print("[FAIL] Frontmatter validation")
        for e in errors:
            print(f"  - {e}")
        all_passed = False
    
    # 3. Meta JSON 验证
    passed, errors = validate_meta_json(skill_path)
    if passed:
        print("[PASS] _meta.json validation")
    else:
        print("[FAIL] _meta.json validation")
        for e in errors:
            print(f"  - {e}")
        all_passed = False
    
    print("-" * 50)
    if all_passed:
        print("All validations passed!")
        sys.exit(0)
    else:
        print("Validation failed.")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: skills/clawhub/zarzoom/scripts/validate-skill.mjs
#!/usr/bin/env node
/**
 * openclaw-skill/scripts/validate-skill.mjs
 *
 * Validates SKILL.md against the ClawHub frontmatter spec before
 * submission. Lint-style — fails fast with a clear message; doesn't
 * try to be comprehensive (ClawHub's server runs the canonical check
 * at publish time).
 *
 * Run via:
 *   node openclaw-skill/scripts/validate-skill.mjs
 *
 * Exits 0 on success, 1 on validation failure.
 *
 * Reference: docs.openclaw.ai/tools/skills + openclaw/clawhub/docs/skill-format.md
 */

import { readFileSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const SKILL_PATH = resolve(__dirname, "..", "SKILL.md");

const errors = [];
const warn = (msg) => console.warn("[warn]", msg);
const fail = (msg) => errors.push(msg);

let raw;
try {
  raw = readFileSync(SKILL_PATH, "utf8");
} catch (err) {
  console.error(`[fatal] Could not read ${SKILL_PATH}:`, err.message);
  process.exit(1);
}

// Extract YAML frontmatter (between the first two `---` lines).
const fmMatch = raw.match(/^---\r?\n([\s\S]*?)\r?\n---\r?\n([\s\S]*)$/);
if (!fmMatch) {
  fail("SKILL.md is missing YAML frontmatter (must start with `---` ... `---`).");
} else {
  const fm = fmMatch[1];
  const body = fmMatch[2];

  // Parse with a minimal hand-rolled YAML reader — we only support
  // single-line key: value pairs + simple lists, which is all the
  // ClawHub spec accepts anyway ("frontmatter keys must be single-line").
  const fields = {};
  let inList = null;
  for (const line of fm.split(/\r?\n/)) {
    const stripped = line.replace(/^\s+/, "");
    if (stripped === "" || stripped.startsWith("#")) continue;

    const listItem = line.match(/^\s+-\s+(.*)$/);
    if (listItem && inList) {
      fields[inList].push(listItem[1].trim());
      continue;
    }
    inList = null;

    const kv = line.match(/^([a-zA-Z_][\w.-]*)\s*:\s*(.*)$/);
    if (!kv) continue;
    const [, key, value] = kv;
    if (value === "") {
      fields[key] = [];
      inList = key;
    } else {
      fields[key] = value.trim().replace(/^["']|["']$/g, "");
    }
  }

  // Required fields per ClawHub skill-format spec.
  if (!fields.name) fail("frontmatter is missing required field `name`.");
  else if (!/^[a-z][a-z0-9-]*$/.test(fields.name)) {
    fail(`frontmatter \`name\` must be lowercase + hyphens (got "${fields.name}").`);
  }

  if (!fields.description) fail("frontmatter is missing required field `description`.");
  else if (fields.description.length < 20) {
    warn(`frontmatter \`description\` is very short (${fields.description.length} chars); aim for ≥ 50 for vector-search ranking.`);
  } else if (fields.description.length > 500) {
    warn(`frontmatter \`description\` is long (${fields.description.length} chars); ClawHub may truncate.`);
  }

  // Optional but strongly recommended.
  if (!fields.version) warn("frontmatter has no `version` (semver recommended).");
  else if (!/^\d+\.\d+\.\d+(-[\w.]+)?$/.test(fields.version)) {
    fail(`frontmatter \`version\` should be semver (got "${fields.version}").`);
  }

  if (!fields.homepage) warn("frontmatter has no `homepage` URL.");
  else if (!/^https?:\/\//.test(fields.homepage)) {
    fail(`frontmatter \`homepage\` must be an http(s) URL (got "${fields.homepage}").`);
  }

  if (!fields.primaryEnv && !fields.requires?.length) {
    warn("frontmatter declares no `primaryEnv` or `requires.env` — is this skill really credential-free?");
  }

  // Body sanity checks.
  if (body.trim().length < 200) {
    fail("SKILL.md body is suspiciously short (<200 chars). The body is what guides the model — it should be substantial.");
  }
}

// Print + exit.
if (errors.length === 0) {
  console.log("[ok] SKILL.md frontmatter validates.");
  process.exit(0);
} else {
  console.error("[fail] SKILL.md validation failed:");
  for (const e of errors) console.error("  -", e);
  process.exit(1);
}
